[FLINK-10068][docs] Add documentation for RocksDB-based timers and st…

StefanRRichter opened the pull request:
## What is the purpose of the change

This PR updates the documentation about timers:
- Stopping/removing timers.
- RocksDB-based timers.
- Asynchronous snapshots for timers.

twalthr commented on docs/dev/stream/operators/process_function.md:
```diff
@@ -277,19 +277,18 @@ Both types of timers (processing-time and event-time) are internally maintained
 
 The `TimerService` deduplicates timers per key and timestamp, i.e., there is at most one timer per key and timestamp. If multiple timers are registered for the same timestamp, the `onTimer()` method will be called just once.
```
Can we move this entire `Fault Tolerance` section to `dev/stream/state/checkpointing.md` for a new `Timers` section? This is a general concept that is not only relevant for `ProcessFunction`.

twalthr replied:
Writing more docs about this might be a bigger change but moving 10 lines into a subsection on a different page can be done in this PR.

twalthr commented on docs/dev/stream/operators/process_function.md:
```diff
@@ -277,19 +277,18 @@ Both types of timers (processing-time and event-time) are internally maintained
-**Note:** Timers are always synchronously checkpointed, regardless of the configuration of the state backends. 
-Therefore, a large number of timers can significantly increase checkpointing time. 
-See the "Timer Coalescing" section for advice on how to reduce the number of timers.
+<span class="label label-info">Note</span> Timers are always asynchronously checkpointed, except for the combination of RocksDB backend / with incremental snapshots / with heap-based timers (will be resolved with `FLINK-10026`).
```
It would be great if you could add more internal knowledge if there is more to say. This section is very tiny but I guess a lot of work went it to so we should document it accordingly.

twalthr replied:
If there is nothing more to say in your opinion, then I'm fine with this.

twalthr commented on docs/dev/stream/operators/process_function.md:
```diff
@@ -333,3 +332,43 @@ ctx.timerService.registerEventTimeTimer(coalescedTime)
+<div data-lang="java" markdown="1">
+{% highlight java %}
+long timestampOfTimerToStop = ...
+ctx.timerService().deleteProcessingTimeTimer(timestampOfTimerToStop);
```
Are there any performance implications of deleting a timer? I guess not anymore right?

twalthr commented on docs/dev/stream/operators/process_function.md:
```diff
@@ -333,3 +332,43 @@ ctx.timerService.registerEventTimeTimer(coalescedTime)
+</div>
+</div>
+
+<span class="label label-info">Note</span> Stopping a timer has no effect if no such timer timer with the given timestamp is registered.
```
remove duplicate `timer`

twalthr commented on docs/ops/state/large_state_tuning.md:
```diff
@@ -142,6 +142,17 @@ by default. To enable this feature, users can instantiate a `RocksDBStateBackend
 
+**RocksDB Timers**
+
+For RocksDB, user can chose whether timers are stored on the heap (default) or inside RocksDB. Heap-based timers can have a better performance for smaller amounts of
```
"a user"

alpinegizmo replied:
Please change "amounts" to "numbers" in both places in this sentence. (This is similar to the distinction in English between less and fewer. Continuous quantities have amounts, while countable quantities have numbers.)

twalthr commented on docs/ops/state/large_state_tuning.md:
```diff
@@ -142,6 +142,17 @@ by default. To enable this feature, users can instantiate a `RocksDBStateBackend
+For RocksDB, user can chose whether timers are stored on the heap (default) or inside RocksDB. Heap-based timers can have a better performance for smaller amounts of
+timers, while storing timers inside RocksDB offers higher scalability as the amount of timers in RocksDB can exceed the available main memory (spilling to disk).
+
+When using RockDB as state backend, the type of timer storage can be selected through Flink's configuration via option key `RocksDBOptions.TIMER_SERVICE_FACTORY`.
```
The option classes are mostly internal, right? So a proper YAML would be more useful I guess. Can I only set this property for an entire Flink cluster or also per-job?

twalthr commented on docs/ops/state/large_state_tuning.md:
```diff
@@ -142,6 +142,17 @@ by default. To enable this feature, users can instantiate a `RocksDBStateBackend
+timers, while storing timers inside RocksDB offers higher scalability as the amount of timers in RocksDB can exceed the available main memory (spilling to disk).
+
+When using RockDB as state backend, the type of timer storage can be selected through Flink's configuration via option key `RocksDBOptions.TIMER_SERVICE_FACTORY`.
+Possible choices are `HEAP` (to store timers on the heap, default) and `ROCKSDB` (to store timers in RocksDB).
```
Usually all YAML values are lower case. These are upper case?

StefanRRichter replied:
They are the names of some enum constants, but we are case-insensitive when evaluating them.

alpinegizmo commented on docs/ops/state/large_state_tuning.md:
```diff
@@ -142,6 +142,17 @@ by default. To enable this feature, users can instantiate a `RocksDBStateBackend
+Possible choices are `HEAP` (to store timers on the heap, default) and `ROCKSDB` (to store timers in RocksDB).
+
+<span class="label label-info">Note</span> *The combination RocksDB state backend / with incremental checkpoint / with heap-based timers currently does NOT support asynchronous snapshots for the timers state.
+Other state like keyed state is still snapshotted asynchronously. Please not that this is no regression from previous versions and will be resolved with `FLINK-10026`.*
```
Please note that this is not a regression ...

alpinegizmo commented on docs/ops/state/large_state_tuning.md:
```diff
@@ -177,12 +188,11 @@ Flink provides some predefined collections of option for RocksDB for different s
 found a set of options that work well and seem representative for certain workloads.
 
-**Important:** RocksDB is a native library, whose allocated memory not from the JVM, but directly from the process'
+<span class="label label-info">Note</span> RocksDB is a native library, whose allocated memory not from the JVM, but directly from the process'
```
RocksDB is a native library that allocates memory directly from the process, and not from the JVM.

alpinegizmo commented on docs/ops/state/large_state_tuning.md:
```diff
@@ -177,12 +188,11 @@ Flink provides some predefined collections of option for RocksDB for different s
+<span class="label label-info">Note</span> RocksDB is a native library, whose allocated memory not from the JVM, but directly from the process'
 native memory. Any memory you assign to RocksDB will have to be accounted for, typically by decreasing the JVM heap size
 of the TaskManagers by the same amount. Not doing that may result in YARN/Mesos/etc terminating the JVM processes for
 allocating more memory than configures.
```
... more memory than configured.

StefanRRichter commented:
@twalthr @alpinegizmo thanks for the review! I addressed all comments from David and some from Timo (see my comments for those that are unchanged for now).
@twalthr please have another look if that is ok for you so that we can provide at least the initial documentation now. We can still add and improve content in another PR.

twalthr commented:
Thanks for the update @StefanRRichter. LGTM % the single comment. Feel free to merge this.

StefanRRichter commented:
Merging.
